File: data_structures/javascript/Trie.js
class TrieNode {
    constructor(key) {
        this.key = key;
        this.parent = null;
        this.prefixes = 0;
        this.children = {};
        this.wordEnd = false;
    }

    //O(word length)
    buildWord() {
        let word = "";
        let currentNode = this;
        while (currentNode) {
            word += currentNode.key;
            currentNode = currentNode.parent;
        }
        return word.reverse();
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode(null);
        this.wordCount = 0;
    }

    //O(word length)
    addWord(word, currentNode = this.root) {
        currentNode.prefixes++;
        let char = word.charAt(0);
        if (!currentNode.children[char]) {
            let newNode = new TrieNode(char);
            currentNode.children[char] = newNode;
        }
        if (word.length > 1) {
            word = word.substring(1);
            currentNode = currentNode.children[char];
            this.addWord(word, currentNode);
        } else {
            currentNode.children[char].wordEnd = true;
            this.wordCount++;
        }
    }

    //O(word length)
    contains(word, currentNode = this.root) {
        let char = word.charAt(0);
        if (!currentNode.children[char]) {
            return false;
        } else {
            if (word.length > 1) {
                word = word.substring(1);
                currentNode = currentNode.children[char];
                return this.contains(word, currentNode);
            } else {
                return currentNode.children[char].wordEnd;
            }
        }
    }

    //O(word length)
    removeWord(word, currentNode = this.root) {
        if (this.contains(word)) {
            // console.log('deleting')
            this.removeUtil(word);
        }
    }

    //O(word length)
    removeUtil(word, currentNode = this.root) {
        currentNode.prefixes--;
        let char = word.charAt(0);
        if (currentNode.children[char]) {
            let remainder = word.substring(1);
            if (remainder) {
                if (currentNode.children[char].prefixes === 1) {
                    delete currentNode.children[char];
                } else {
                    currentNode = currentNode.children[char];
                    this.removeUtil(remainder, currentNode);
                }
            } else {
                if (currentNode.children[char].prefixes === 0) {
                    delete currentNode.children[char];
                } else {
                    currentNode.wordEnd = false;
                    this.wordCount--;
                }
            }
        }
    }

    searchByPrefix(prefix) {
        let currentNode = this.root;
        for (let i = 0; i < prefix.length; i++) {
            let char = prefix.charAt(i);
            if (currentNode.children[char]) {
                currentNode = currentNode[char];
            } else {
                return [];
            }
        }
        let results = [];
        this.retrieveAllWithPrefix(prefix, results);
        return results.reverse();
    }

    retrieveAllWithPrefix(prefixEndNode, results) {
        if (prefixEndNode.wordEnd) {
            results.push(prefixEndNode.buildWord());
        } else {
            for (let child in prefixEndNode.children) {
                this.retrieveAllWithPrefix(child, results);
            }
        }
    }

    print() {
        const newLine = new TrieNode('|');
        const processingQueue = [this.root, newLine];
        let resultString = "";
        while (processingQueue.length) {
            let currentNode = processingQueue.shift();
            if (currentNode.key) {
                resultString += currentNode.key + ' ';
            }
            if (currentNode.key === "|" && processingQueue.length) {
                processingQueue.push(newLine);
            }
            for (let child in currentNode.children) {
                processingQueue.push(currentNode.children[child]);
            }
        }
        console.log(resultString);
    }

    printByDepth() {
        const newLine = new TrieNode('\n');
        const processingQueue = [this.root, newLine];
        let resultString = "";
        while (processingQueue.length) {
            let currentNode = processingQueue.shift();
            if (currentNode.key) {
                resultString += currentNode.key + (currentNode.key === '\n' ? ' ' : '');
            }
            if (currentNode.key === "\n" && processingQueue.length) {
                processingQueue.push(newLine);
            }
            for (let child in currentNode.children) {
                processingQueue.push(currentNode.children[child]);
            }
        }
    }
}

File: data_structures/ruby/trie.rb
class TrieNode
    attr_accessor :key, :parent, :prefixes, :children, :word_end?
    def initialize(key)
        @key = key
        @parent = nil
        @prefixes = 0
        @children = {}
        @word_end? = false
    end

    def build_word
        word = ""
        current_node = self
        until current_node.nil?
            word += current_node.key
            current_node = current_node.parent
        end
        return word.reverse
    end
end

class Trie
    def initialize
        @root = TrieNode.new(nil)
        @word_count = 0
    end

    def add_word(word, current_node = @root)
        current_node.prefixes += 1
        char = word[0]
        if !current_node.children.has_key(char)
            new_node = TrieNode.new(char)
            current_node.children[char] = new_node
        end
        if word.length > 1
            word = word.drop(1)
            current_node = current_node[char]
            add_word(word, current_node)
        else
            current_node.children[char].word_end? = true
            @word_count += 1
        end
    end

    def contains(word, current_node = @root)
        char = word[0]
        if !current_node.children.has_key(char)
            return false
        else
            if (word.length > 1)
                word = word.drop(1)
                current_node = current_node.children[char]
                return contains(word, current_node)
            else
                return current_node.children[char].word_end?
            end
        end
    end

    def remove_word(word, current_node = @root)
        if (contains(word))
            remove_helper(word)
        end

        def remove_helper(word, current_node = @root)
            current_node.prefixes -= 1;
            char = word[0]
            if current_node.children.has_key(char)
                if word.length > 1
                    if current_node.children[char].prefixes == 1
                        current_node.children.delete(char)
                        @word_count -= 1
                    else
                        current_node = current_node.children[char]
                        remove_helper(word.drop(1), current_node)
                    end
                else
                    if current_node.children[char].prefixes == 0
                        current_node.children.delete(char)
                    else
                        current_node.word_end = false
                        @word_count -= 1
                    end
                end
            else
            end
        end
    end

    def search_by_prefix(prefix)
        current_node = @root
        prefix.each_char do |char|
            if current_node.children[char]
                current_node = current_node[char]
            else
                return []
            end
        end
        search_results = []
        retrieve_all_with_prefix(current_node, search_results)
        return search_results.reverse
    end

    def retrieve_all_with_prefix(prefix_end_node, results)
        if prefix_end_node.word_end?
            results.push(prefix_end_node.build_word)
        else
            prefix_end_node.children.each do |child|
                retrieve_all_with_prefix(child, results)
            end
        end
    end

    def print_trie
        new_line = TrieNode.new('|')
        processing_queue = [@root, new_line]
        result_string = ""

        until processing_queue.length == 0
            current_node = processing_queue.shift 
            if current_node.key
                result_string += current_node.key + " "
            end
            if current_node.key == '|' && processing_queue.length > 0
                processing_queue.push(new_line)
            end
            if current_node.children.size > 0
                current_node.children.keys.each do |child|
                    processing_queue.push(current_node[child])
                end
            end
        end

        return result_string
    end

    def print_trie_by_depth 
        new_line = TrieNode.new("\n")
        processing_queue = [@root, new_line]
        result_string = ""

        until processing_queue.empty?
            current_node = processing_queue.shift 
            if current_node.key
                result_string += current_node.key + (current_node.key == "\n" ? " " : "")
            end
            if (current_node.key == "\n" && !processing_queue.empty?)
                processing_queue.push(new_line)
            end
            if current_node.children.size > 0
                current_node.children.keys.each do |child|
                    processing_queue.push(current_node.children[child])
                end
            end
        end
    end
end